lib/SILOptimizer/IPO/ClosureSpecializer.cpp

```diff
@@ -1255,6 +1302,60 @@ bool SILClosureSpecializerTransform::gatherCallSites(
           continue;
         }
 
+        // Make sure the closure callee function and the apply callee function
```

Comment:
I don't understand what's going on below. It sounds like the transformation is not general enough to handle all combinations of the caller's and callee generic signature. It should be pretty easy to build the right signature though, using a new GenericSignatureBuilder.

Reply:
`GenericSignatureBuilder` may fix a lot of complexity here. 

Here's the main problem. Sometimes the function called (to which the closure is passed) has a different set of generics than the function in which that function is called (the caller). Sometimes the callee takes a `partial_apply` as a closure argument. The `partial_apply` may have generics in its substitution map that **don't** exist in the callee. But, this pass will still try to pull the `partial_apply` out of the caller and into the (cloned) callee. This won't work because those generics don't exist in the callee. So, we check that all generics in the `parial_apply`'s substitution map also exist in the "normal" apply's substitution map.